File: plugins/openai-apps-sdk/skills/mcp-server-architecture/examples/minimal-server.py
#!/usr/bin/env python3
"""Minimal MCP server for OpenAI Apps SDK.

This example demonstrates the basic structure of an MCP server
that can be used with ChatGPT apps.

Usage:
    pip install mcp
    python minimal-server.py

Then expose via ngrok:
    ngrok http 8000
"""

from mcp.server.fastmcp import FastMCP

# Create the MCP server
mcp = FastMCP("minimal-server")


@mcp.tool()
def hello(name: str) -> dict:
    """Say hello to someone.

    Args:
        name: The name of the person to greet
    """
    return {
        "structuredContent": {
            "greeting": f"Hello, {name}!",
            "name": name
        },
        "content": [
            {"type": "text", "text": f"Hello, {name}!"}
        ]
    }


@mcp.tool()
def add_numbers(a: float, b: float) -> dict:
    """Add two numbers together.

    Args:
        a: First number
        b: Second number
    """
    result = a + b
    return {
        "structuredContent": {
            "result": result,
            "operation": "addition",
            "operands": [a, b]
        },
        "content": [
            {"type": "text", "text": f"{a} + {b} = {result}"}
        ]
    }


@mcp.resource("ui://widget/main.html")
def main_widget() -> str:
    """Serve a simple widget that displays tool output."""
    return """<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <style>
        body {
            font-family: system-ui, sans-serif;
            padding: 16px;
            margin: 0;
        }
        .result {
            background: #f0f0f0;
            padding: 12px;
            border-radius: 8px;
        }
    </style>
</head>
<body>
    <div class="result" id="output">Loading...</div>
    <script>
        const data = window.openai?.toolOutput?.structuredContent;
        if (data) {
            document.getElementById('output').textContent = JSON.stringify(data, null, 2);
        }
    </script>
</body>
</html>"""


if __name__ == "__main__":
    print("Starting MCP server on http://localhost:8000/mcp")
    print("Expose with: ngrok http 8000")
    mcp.run(transport="streamable-http", host="0.0.0.0", port=8000)

File: plugins/openai-apps-sdk/skills/mcp-server-architecture/examples/minimal-server.ts
/**
 * Minimal MCP server for OpenAI Apps SDK (TypeScript)
 *
 * This example demonstrates the basic structure of an MCP server
 * that can be used with ChatGPT apps.
 *
 * Usage:
 *   npm install @modelcontextprotocol/sdk zod express
 *   npx tsx minimal-server.ts
 *
 * Then expose via ngrok:
 *   ngrok http 8000
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  ListResourcesRequestSchema,
  ReadResourceRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";
import express from "express";
import { randomUUID } from "crypto";

// Create the MCP server
const server = new Server(
  { name: "minimal-server", version: "1.0.0" },
  { capabilities: { tools: {}, resources: {} } }
);

// List available tools
server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [
    {
      name: "hello",
      description: "Say hello to someone",
      inputSchema: {
        type: "object",
        properties: {
          name: { type: "string", description: "The name of the person to greet" }
        },
        required: ["name"]
      }
    },
    {
      name: "add_numbers",
      description: "Add two numbers together",
      inputSchema: {
        type: "object",
        properties: {
          a: { type: "number", description: "First number" },
          b: { type: "number", description: "Second number" }
        },
        required: ["a", "b"]
      }
    }
  ]
}));

// Handle tool calls
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  switch (name) {
    case "hello": {
      const greeting = `Hello, ${args.name}!`;
      return {
        structuredContent: {
          greeting,
          name: args.name
        },
        content: [{ type: "text", text: greeting }]
      };
    }

    case "add_numbers": {
      const result = args.a + args.b;
      return {
        structuredContent: {
          result,
          operation: "addition",
          operands: [args.a, args.b]
        },
        content: [{ type: "text", text: `${args.a} + ${args.b} = ${result}` }]
      };
    }

    default:
      throw new Error(`Unknown tool: ${name}`);
  }
});

// List available resources
server.setRequestHandler(ListResourcesRequestSchema, async () => ({
  resources: [
    {
      uri: "ui://widget/main.html",
      name: "Main Widget",
      mimeType: "text/html+skybridge"
    }
  ]
}));

// Serve resources
server.setRequestHandler(ReadResourceRequestSchema, async (request) => {
  const { uri } = request.params;

  if (uri === "ui://widget/main.html") {
    return {
      contents: [{
        uri,
        mimeType: "text/html+skybridge",
        text: `<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <style>
        body {
            font-family: system-ui, sans-serif;
            padding: 16px;
            margin: 0;
        }
        .result {
            background: #f0f0f0;
            padding: 12px;
            border-radius: 8px;
        }
    </style>
</head>
<body>
    <div class="result" id="output">Loading...</div>
    <script>
        const data = window.openai?.toolOutput?.structuredContent;
        if (data) {
            document.getElementById('output').textContent = JSON.stringify(data, null, 2);
        }
    </script>
</body>
</html>`
      }]
    };
  }

  throw new Error(`Unknown resource: ${uri}`);
});

// Setup Express and transport
const app = express();

const transport = new StreamableHTTPServerTransport({
  sessionIdGenerator: () => randomUUID()
});

app.use("/mcp", async (req, res) => {
  await transport.handleRequest(req, res, server);
});

// Connect and start
await server.connect(transport);

const PORT = process.env.PORT || 8000;
app.listen(PORT, () => {
  console.log(`MCP server running on http://localhost:${PORT}/mcp`);
  console.log("Expose with: ngrok http " + PORT);
});
